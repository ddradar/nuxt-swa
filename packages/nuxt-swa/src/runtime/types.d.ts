interface Claim {
  typ: string
  val: string
}
/**
 * User-identifiable information to your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/user-information?tabs=javascript#client-principal-data
 */
export interface ClientPrincipal {
  /** The name of the identity provider. */
  identityProvider: IdentityProvider
  /**
   * An Azure Static Web Apps-specific unique identifier for the user.
   * - The value is unique on a per-app basis. For instance, the same user returns a different `userId` value on a different Static Web Apps resource.
   * - The value persists for the lifetime of a user. If you delete and add the same user back to the app, a new `userId` is generated.
   */
  userId: string
  /**
   * Username or email address of the user.
   * Some providers return the user's email address, while others send the user handle.
   */
  userDetails: string
  /** An array of the user's assigned roles. */
  userRoles: UserRole[]
  /**
   * An array of claims returned by your custom authentication provider.
   * Only accessible in the direct-access endpoint (/.auth/me).
   */
  claims?: Claim[]
}
