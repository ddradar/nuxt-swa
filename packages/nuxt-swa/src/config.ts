import type { NuxtConfig } from 'nuxt/config'

export const preConfiguredProviders = ['aad' as const, 'github' as const]
const knownIdentityProviders = [
  'apple',
  'facebook',
  'google',
  'twitter',
] as const

export interface ModuleOptions {
  /**
   * Authentication Provider list to login your app.
   * If your app does not use built-in Authentication/Authorization, please set this to `[]`.
   * @default ['aad', 'github']
   */
  authProviders?: (
    | (typeof knownIdentityProviders)[number]
    | Omit<string, (typeof knownIdentityProviders)[number]>
  )[]
  /**
   * Define custom roles to use in your app besides the default roles (`anonymous`, `authenticated`).
   * @default []
   */
  customRoles: Omit<string, 'anonymous' | 'authenticated'>[]
  /** Data API config */
  dataApi:
    | {
        /** REST endpoint path (same value as `runtime.rest.path` in `staticwebapp.database.config.json`) */ rest: string
        /** GraphQL endpoint path (same value as `runtime.graphql.path` in `staticwebapp.database.config.json`) */ graphql: string
      }
    | false
}
export const defaults: ModuleOptions = {
  customRoles: [],
  dataApi: { rest: '/rest', graphql: '/graphql' },
}

/** Resolve `authProviders` option from Nitro config. */
export function resolveAuthProviders(
  options: Required<Required<NuxtConfig>['nitro']>['azure']['config']
): Required<ModuleOptions>['authProviders'] {
  if (!options?.auth?.identityProviders) return preConfiguredProviders

  const identityProviders = options?.auth?.identityProviders

  const result: Required<ModuleOptions>['authProviders'] = []

  addPreConfiguredProviderIfEnabled('aad') // Microsoft Entra ID (formerly Azure Active Directory)
  addCustomProviderIfEnabled('apple') // Apple
  addCustomProviderIfEnabled('facebook') // Facebook
  addPreConfiguredProviderIfEnabled('github') // GitHub
  addCustomProviderIfEnabled('google') // Google
  addCustomProviderIfEnabled('twitter') // X (formerly Twitter)

  // OpenID Connect
  for (const key in identityProviders.customOpenIdConnectProviders) {
    if (identityProviders.customOpenIdConnectProviders[key]?.enabled !== false)
      result.push(key)
  }

  return result

  function addPreConfiguredProviderIfEnabled(
    provider: (typeof preConfiguredProviders)[number]
  ) {
    const key = provider === 'aad' ? 'azureActiveDirectory' : provider
    if (identityProviders[key]?.enabled !== false) result.push(provider)
  }

  function addCustomProviderIfEnabled(
    provider: (typeof knownIdentityProviders)[number]
  ) {
    if (
      identityProviders[provider] &&
      identityProviders[provider]?.enabled !== false
    )
      result.push(provider)
  }
}

export interface ModulePublicRuntimeConfig {
  /** Nuxt SWA config */
  swa: {
    /** REST API endpoint */ rest: string
    /** GraphQL endpoint */ graphql: string
  }
}

type SpecifiedRole = Exclude<UserRole, 'anonymous'>

declare module '#app' {
  interface PageMeta {
    /** List of roles required to view this page */
    allowedRoles?: SpecifiedRole | SpecifiedRole[]
  }
}

declare module 'vue-router' {
  interface RouteMeta {
    /** List of roles required to view this page */
    allowedRoles?: SpecifiedRole | SpecifiedRole[]
  }
}
