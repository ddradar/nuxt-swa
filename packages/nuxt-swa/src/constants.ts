// Azure Static Web Apps
export const _authBasePath = '/.auth'
export const _clientPrincipalHeader = 'x-ms-client-principal'

// My App
export const _authDataKey = 'swa-principal'
