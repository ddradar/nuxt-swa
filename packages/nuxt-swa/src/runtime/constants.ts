// Azure Static Web Apps
export const _authBasePath = '/.auth'
export const _dataApiBasePath = '/data-api'
export const _clientPrincipalCookie = 'StaticWebAppsAuthCookie'
export const _clientPrincipalHeader = 'x-ms-client-principal'
export const _preConfiguredProviders = ['aad' as const, 'github' as const]
export const _knownIdentityProviders = [
  'apple',
  'facebook',
  'google',
  'twitter',
] as const

// My App
export const _authDataKey = 'swa-principal'
