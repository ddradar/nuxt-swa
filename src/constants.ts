export const _authBasePath = '/.auth'
export const _clientPrincipalHeader = 'x-ms-client-principal'
