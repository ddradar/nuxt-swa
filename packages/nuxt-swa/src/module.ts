import {
  defineNuxtModule,
  createResolver,
  addTypeTemplate,
  addImports,
} from '@nuxt/kit'

import { defaults, resolveAuthProviders } from './config'

export default defineNuxtModule({
  meta: {
    name: 'nuxt-swa',
    configKey: 'swa',
    compatibility: {
      nuxt: '^3.9.0',
    },
  },
  defaults,
  setup(options, nuxt) {
    const resolver = createResolver(import.meta.url)

    if (options.authProviders === undefined)
      options.authProviders = resolveAuthProviders(
        nuxt.options.nitro?.azure?.config
      )

    addTypeTemplate({
      filename: 'types/nuxt-swa.d.ts',
      src: resolver.resolve('runtime/types/nuxt-swa.d.ts.template'),
      options,
    })
    if (options.authProviders.length) {
      // Use auth feature
      addImports({
        name: 'useEasyAuth',
        from: resolver.resolve('runtime/composables/useEasyAuth'),
      })
    }
    // Add `nitro.azure.config` Typing
    addTypeTemplate({
      filename: 'types/AzureOptions.d.ts',
      src: resolver.resolve('runtime/types/AzureOptions.d.ts'),
    })
  },
})
