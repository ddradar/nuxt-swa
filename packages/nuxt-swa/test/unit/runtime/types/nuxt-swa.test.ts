// @vitest-environment node
import { readFile } from 'node:fs/promises'
import { resolve } from 'node:path'

import { template } from 'lodash'
import { describe, expect, it } from 'vitest'

import { defaults, type SWAModuleOptions } from '~/src/config'
import { _preConfiguredProviders } from '~/src/runtime/constants'

describe('runtime/types/nuxt-swa.d.ts.template', async () => {
  const minimumOptions: SWAModuleOptions = {
    authProviders: [],
    customRoles: [],
  }
  const defaultOptions = { ...defaults, authProviders: _preConfiguredProviders }
  const fullOptions: SWAModuleOptions = {
    authProviders: ['github', 'twitter', 'line'],
    customRoles: ['administrator'],
  }
  const templateString = await readFile(
    resolve(
      __dirname,
      '../../../../',
      'src/runtime/types',
      'nuxt-swa.d.ts.template'
    ),
    { encoding: 'utf-8' }
  )

  it.each([defaultOptions, minimumOptions, fullOptions])(
    '%o renders expected Type Definition',
    options => expect(template(templateString)({ options })).toMatchSnapshot()
  )
})
