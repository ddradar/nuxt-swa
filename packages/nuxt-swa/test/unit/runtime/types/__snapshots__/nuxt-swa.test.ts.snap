// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime/types/nuxt-swa.d.ts.template > { authProviders: [ 'github', 'twitter', 'line' ], customRoles: [ 'administrator' ] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa

declare module '#app' {
  interface PageMeta {
    /** List of roles required to view this page */
    allowedRoles?:
      | Exclude<UserRole, 'anonymous'>
      | Exclude<UserRole, 'anonymous'>[]
  }
}

/**
 * Authentication Provider list to login your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#configure-a-custom-identity-provider
 */
type IdentityProvider =
  | 'github'
  | 'twitter'
  | 'line'

/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole =
  | 'anonymous'
  | 'authenticated'
  | 'administrator'

export {}
"
`;

exports[`runtime/types/nuxt-swa.d.ts.template > { authProviders: [], customRoles: [] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa

/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole =
  | 'anonymous'
  | 'authenticated'

export {}
"
`;

exports[`runtime/types/nuxt-swa.d.ts.template > { customRoles: [], authProviders: [ 'aad', 'github' ] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa

declare module '#app' {
  interface PageMeta {
    /** List of roles required to view this page */
    allowedRoles?:
      | Exclude<UserRole, 'anonymous'>
      | Exclude<UserRole, 'anonymous'>[]
  }
}

/**
 * Authentication Provider list to login your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#configure-a-custom-identity-provider
 */
type IdentityProvider =
  | 'aad'
  | 'github'

/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole =
  | 'anonymous'
  | 'authenticated'

export {}
"
`;
